File: src/Observability/InvestigationService.php
<?php
declare(strict_types=1);
namespace App\Observability;
use App\Infrastructure\Database;
use App\Billing\BillingError;

/** A durable support workspace; notes never change billing state. */
final class InvestigationService
{
    public function __construct(private Database $db) {}
    public function start(string $type,string $id,string $title,string $actor): array
    {
        if(!in_array($type,['user','subscription','payment','incident'],true)||$id==='')throw new BillingError('Некорректный объект расследования.');
        $existing=$this->db->one("SELECT * FROM investigations WHERE subject_type=? AND subject_id=? AND status='open'",[$type,$id]); if($existing)return $existing;
        $now=time();$caseId=Database::id();$code='INV-'.date('Y').'-'.random_int(10000,99999);
        $this->db->execute("INSERT INTO investigations(id,code,status,subject_type,subject_id,title,created_by,created_at) VALUES(?,?, 'open',?,?,?,?,?)",[$caseId,$code,$type,$id,mb_substr(trim($title)?:'Расследование',0,200),$actor,$now]);
        return $this->db->one('SELECT * FROM investigations WHERE id=?',[$caseId]);
    }
    public function detail(string $id): ?array
    {
        $case=$this->db->one('SELECT * FROM investigations WHERE id=?',[$id]);if(!$case)return null;
        $case['notes']=$this->db->all('SELECT n.*,u.email,u.telegram_id FROM investigation_notes n LEFT JOIN users u ON u.id=n.author_id WHERE n.investigation_id=? ORDER BY n.created_at',[$id]);
        $case['context']=$this->context($case['subject_type'],$case['subject_id']);return $case;
    }
    public function note(string $caseId,string $body,string $actor): void
    {
        $body=trim($body);if(mb_strlen($body)<2||mb_strlen($body)>2000)throw new BillingError('Заметка должна содержать от 2 до 2000 символов.');
        if(!$this->db->one("SELECT id FROM investigations WHERE id=? AND status='open'",[$caseId]))throw new BillingError('Расследование закрыто или не найдено.');
        $this->db->execute('INSERT INTO investigation_notes(id,investigation_id,author_id,body,created_at) VALUES(?,?,?,?,?)',[Database::id(),$caseId,$actor,$body,time()]);
    }
    public function resolve(string $caseId,string $actor): void
    {
        $this->db->execute("UPDATE investigations SET status='resolved',resolved_at=? WHERE id=? AND status='open'",[time(),$caseId]);
        $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)',[Database::id(),$actor,'investigation.resolved',$caseId,time()]);
    }
    public function list(): array {return $this->db->all('SELECT * FROM investigations ORDER BY CASE status WHEN \'open\' THEN 0 ELSE 1 END,created_at DESC LIMIT 100');}
    public function expectedActual(string $subscriptionId): ?array
    {
        $s=$this->db->one('SELECT s.*,COALESCE(p.name,o.plan_name) plan_name,pa.state,pa.last_synced_at,pa.last_error,pa.external_user_id FROM subscriptions s LEFT JOIN plans p ON p.id=s.plan_id LEFT JOIN orders o ON o.id=s.order_id LEFT JOIN provisioning_accounts pa ON pa.subscription_id=s.id WHERE s.id=?',[$subscriptionId]);if(!$s)return null;
        $expected=['Статус подписки'=>$s['lifecycle_status']??$s['status'],'Срок действия'=>gmdate('d.m.Y H:i',(int)$s['expires_at']),'Лимит трафика'=>((int)($s['traffic_limit_gb']??0)===0?'Безлимит':$s['traffic_limit_gb'].' ГБ'),'Выдача VPN'=>'active'];
        $actual=['Статус подписки'=>$s['status'],'Срок действия'=>gmdate('d.m.Y H:i',(int)$s['expires_at']),'Лимит трафика'=>((int)($s['traffic_limit_gb']??0)===0?'Безлимит':$s['traffic_limit_gb'].' ГБ'),'Выдача VPN'=>$s['state']??'не создана'];
        $rows=[];foreach($expected as $name=>$want)$rows[]=['name'=>$name,'expected'=>$want,'actual'=>$actual[$name],'ok'=>$want===$actual[$name]||($name==='Выдача VPN'&&$actual[$name]==='active')];
        return ['subscription'=>$s,'rows'=>$rows,'freshness'=>$s['last_synced_at']?max(0,time()-(int)$s['last_synced_at']):null];
    }
    public function whyNotRenewed(string $subscriptionId): array
    {
        $s=$this->db->one('SELECT s.*,u.balance_kopeks,COALESCE(p.price_minor,o.price_minor) price_minor FROM subscriptions s JOIN users u ON u.id=s.user_id LEFT JOIN plans p ON p.id=COALESCE(s.renew_plan_id,s.plan_id) LEFT JOIN orders o ON o.id=s.order_id WHERE s.id=?',[$subscriptionId]);if(!$s)throw new BillingError('Подписка не найдена.');
        $reasons=[];$now=time();
        if(!(int)($s['auto_renew']??0))$reasons[]=['ok'=>false,'text'=>'Автопродление выключено пользователем или оператором.'];
        elseif((int)($s['expires_at']??0)>$now)$reasons[]=['ok'=>false,'text'=>'Срок продления ещё не наступил.'];
        elseif((int)$s['balance_kopeks']<(int)$s['price_minor'])$reasons[]=['ok'=>false,'text'=>'Недостаточно средств: доступно '.number_format((int)$s['balance_kopeks']/100,2,',',' ').' ₽, требуется '.number_format((int)$s['price_minor']/100,2,',',' ').' ₽.'];
        elseif($s['renew_order_id'])$reasons[]=['ok'=>true,'text'=>'Заказ на продление уже создан и ожидает обработку.'];
        else $reasons[]=['ok'=>false,'text'=>'Не найдено автоматического заказа: требуется проверка планировщика.'];
        return ['subscription'=>$s,'reasons'=>$reasons];
    }
    public function smartQueues(): array
    {
        return ['need_attention'=>(int)($this->db->one("SELECT COUNT(*) c FROM provisioning_accounts WHERE state IN ('retry','failed')")['c']??0),'payment_issues'=>(int)($this->db->one("SELECT COUNT(*) c FROM payments WHERE status IN ('pending','failed')")['c']??0),'provisioning'=>(int)($this->db->one("SELECT COUNT(*) c FROM provisioning_accounts WHERE state IN ('pending','processing','retry')")['c']??0),'manual_review'=>(int)($this->db->one("SELECT COUNT(*) c FROM operational_cases WHERE status='open'")['c']??0)];
    }
    private function context(string $type,string $id): array
    {
        return match($type){'subscription'=>$this->expectedActual($id)??[],'user'=>['user'=>$this->db->one('SELECT id,email,telegram_id FROM users WHERE id=?',[$id])],'payment'=>['payment'=>$this->db->one('SELECT * FROM payments WHERE id=?',[$id])],'incident'=>['incident'=>$this->db->one('SELECT * FROM incidents WHERE id=?',[$id])],default=>[]};
    }
}

File: rails_backend/app/services/operations/investigation_service.rb
module Operations
  class InvestigationService
    SUBJECT_TYPES = %w[user subscription payment incident].freeze

    def list
      Investigation.order(Arel.sql("CASE status WHEN 'open' THEN 0 ELSE 1 END"), created_at: :desc).limit(100)
        .as_json(include: { investigation_notes: { only: %i[id author_id body created_at] } })
    end

    def detail(id)
      record = Investigation.find_by(id: id)
      return unless record

      data = record.as_json
      data["notes"] = InvestigationNote.where(investigation_id: record.id).joins("LEFT JOIN users ON users.id = investigation_notes.author_id")
        .order(created_at: :asc).pluck("investigation_notes.id", "investigation_notes.author_id", "investigation_notes.body",
          "investigation_notes.created_at", "users.email", "users.telegram_id").map do |note_id, author_id, body, created_at, email, telegram_id|
          { id: note_id, author_id: author_id, body: body, created_at: created_at, email: email, telegram_id: telegram_id }
        end
      data["context"] = context(record.subject_type, record.subject_id)
      data
    end

    def start(type:, subject_id:, title:, actor:)
      type = type.to_s
      subject_id = subject_id.to_s.strip
      raise Billing::BillingError, "Некорректный объект расследования." unless SUBJECT_TYPES.include?(type) && subject_id.present? && subject_exists?(type, subject_id)

      existing = Investigation.find_by(subject_type: type, subject_id: subject_id, status: "open")
      return existing if existing

      now = Time.now.to_i
      Investigation.create!(id: Infrastructure::IdGenerator.call,
        code: "INV-#{Time.now.utc.year}-#{SecureRandom.random_number(90_000) + 10_000}",
        status: "open", subject_type: type, subject_id: subject_id,
        title: title.to_s.strip.presence&.first(200) || "Расследование",
        created_by: actor, created_at: now)
    rescue ActiveRecord::RecordNotUnique
      Investigation.find_by!(subject_type: type, subject_id: subject_id, status: "open")
    end

    def note(case_id:, body:, actor:)
      body = body.to_s.strip
      raise Billing::BillingError, "Заметка должна содержать от 2 до 2000 символов." unless body.length.between?(2, 2000)

      ApplicationRecord.transaction do
        record = Investigation.lock.find_by(id: case_id, status: "open")
        raise Billing::BillingError, "Расследование закрыто или не найдено." unless record
        InvestigationNote.create!(id: Infrastructure::IdGenerator.call, investigation_id: record.id,
          author_id: actor, body: body, created_at: Time.now.to_i)
      end
    end

    def resolve(case_id:, actor:)
      ApplicationRecord.transaction do
        record = Investigation.lock.find_by(id: case_id, status: "open")
        raise Billing::BillingError, "Расследование закрыто или не найдено." unless record
        now = Time.now.to_i
        record.update!(status: "resolved", resolved_at: now)
        AuditLog.create!(id: Infrastructure::IdGenerator.call, actor: actor,
          action: "investigation.resolved", subject: record.id, created_at: now)
      end
    end

    def smart_queues
      { need_attention: ProvisioningAccount.where(state: %w[retry failed]).count,
        payment_issues: Payment.where(status: %w[pending failed]).count,
        provisioning: ProvisioningAccount.where(state: %w[pending processing retry]).count,
        manual_review: OperationalCase.where(status: "open").count }
    end

    def expected_actual(subscription_id)
      subscription = Subscription.includes(:user).find_by(id: subscription_id)
      return unless subscription

      account = ProvisioningAccount.find_by(subscription_id: subscription.id)
      remote = nil
      error = "Сверка с Remnawave не настроена"
      config = Infrastructure::RuntimeConfig.new
      if config.fetch("REMNAWAVE_URL", "").present? && config.fetch("REMNAWAVE_TOKEN", "").present?
        begin
          remote = Integrations::RemnawaveClient.new.resolve(subscription.attributes.symbolize_keys)
          error = "Пользователь не найден в Remnawave" unless remote
        rescue StandardError
          error = "Не удалось получить текущие данные Remnawave"
        end
      end

      expiry = remote && Time.iso8601(remote["expireAt"].to_s).to_i rescue nil
      expected_bytes = subscription.traffic_limit_gb.to_i.zero? ? 0 :
        (subscription.traffic_limit_gb.to_i + subscription.purchased_traffic_gb.to_i) * 1_073_741_824
      actual_bytes = remote && remote["trafficLimitBytes"]&.to_i
      rows = [
        { name: "Срок действия", expected: Time.at(subscription.expires_at.to_i).utc.strftime("%d.%m.%Y %H:%M UTC"),
          actual: expiry ? Time.at(expiry).utc.strftime("%d.%m.%Y %H:%M UTC") : "Не проверено",
          ok: expiry.present? && (expiry - subscription.expires_at.to_i).abs <= 60 },
        { name: "Лимит трафика", expected: expected_bytes.zero? ? "Безлимит" : "#{(expected_bytes / 1_073_741_824.0).round} ГБ",
          actual: actual_bytes.nil? ? "Не проверено" : (actual_bytes.zero? ? "Безлимит" : "#{(actual_bytes / 1_073_741_824.0).round} ГБ"),
          ok: !actual_bytes.nil? && actual_bytes == expected_bytes },
        { name: "Статус VPN", expected: "ACTIVE", actual: remote&.dig("status") || "Не проверено",
          ok: remote&.dig("status") == "ACTIVE" }
      ]
      { subscription: subscription.as_json(include: :user), rows: rows, freshness: remote ? 0 : nil, error: remote ? nil : error }
    end

    def why_not_renewed(subscription_id)
      subscription = Subscription.includes(:user).find_by(id: subscription_id)
      raise Billing::BillingError, "Подписка не найдена." unless subscription
      plan = Plan.find_by(id: subscription.renew_plan_id.presence || subscription.plan_id)
      price_minor = plan&.price_minor.to_i
      balance = subscription.user.balance_kopeks.to_i
      reasons = if subscription.auto_renew.to_i.zero?
        [{ ok: false, text: "Автопродление выключено пользователем или оператором." }]
      elsif subscription.expires_at.to_i > Time.now.to_i
        [{ ok: false, text: "Срок продления ещё не наступил." }]
      elsif balance < price_minor
        [{ ok: false, text: "Недостаточно средств: доступно #{format_minor(balance)}, требуется #{format_minor(price_minor)}." }]
      elsif subscription.renew_order_id.present?
        [{ ok: true, text: "Заказ на продление уже создан и ожидает обработку." }]
      else
        [{ ok: false, text: "Не найдено автоматического заказа: требуется проверка планировщика." }]
      end
      { subscription: subscription.as_json, reasons: reasons }
    end

    private

    def subject_exists?(type, id)
      { "user" => User, "subscription" => Subscription, "payment" => Payment, "incident" => Incident }.fetch(type).exists?(id: id)
    end

    def context(type, id)
      case type
      when "user" then { user: User.where(id: id).pick(:id, :email, :telegram_id) }
      when "subscription" then { subscription: expected_actual(id) }
      when "payment" then { payment: Payment.find_by(id: id)&.as_json }
      when "incident" then { incident: Incident.find_by(id: id)&.as_json }
      else {}
      end
    end

    def format_minor(amount)
      "#{format('%.2f', amount.to_i / 100.0).tr('.', ',')} ₽"
    end
  end
end
